pkg/asn1/aper/marshal.go

```diff
@@ -346,6 +346,187 @@ func (pd *perRawBitData) appendBool(value bool) (err error) {
+
+	if value < 0 {
+		value = -value
+	}
+
+	bitAmount := howManyBitsNeeded(value)
+
+	for {
+		if bitAmount > 8 {
+			bitAmount = bitAmount - 8
+			byteAmount++
+		} else {
+			break
+		}
+	}
+	byteAmount++
+
+	return
+}
+
+// it looks like encoding of REAL doesn't take into account constraints at all, so we don't bother about parsing constraints
+// general rules are - mantissa should be an odd number or a 0
+func (pd *perRawBitData) appendReal(value float64) (err error) {
+
+	log.Debugf("Encoding REAL number %v", value)
+	var mantissa int64
+	var exponent int64
+	var p int
+	var n int
+	negativeExponent := false
+
+	// First, checking whether we encode a whole number
+	if value == math.Trunc(value) {
+		log.Debugf("We're encoding a whole number")
+		// Valid for whole numbers: divide value on 2 until the result is odd, once result is even, stop.
+		// Encode power of 2 (obtained from division) as exponent and encode result of division as a mantissa
+		// If the number is even at the beginning, then encode 0 as an exponent and encode
+		// number as a mantissa (don't forget to put 00 octet in the beginning, for some reason I don't understand (yet))
+		mantissa = int64(value)
+		// mantissa can't be negative
+		if mantissa < 0 {
+			mantissa = -mantissa
+		}
+		for {
+			if mantissa%2 != 0 {
+				break
+			}
+			exponent++
+			mantissa = mantissa / 2
+		}
+		log.Debugf("Obtained mantissa is %v, exponent is %v", mantissa, exponent)
+	} else {
+		log.Debugf("We're encoding a number with a floating point")
+		// For values with numbers after decimal dot (radix), representation is different
+		// Steps are following: multiply initial number by two until it becomes a whole number
+		// if the number is not becoming a whole one, then multiply by 2 (max. 51 times), then
+		// take the resulting number and encode it the same way as a whole number (exponent is encoded as its 2's complement)
+
+		val := value
+		for i := 0; i < 52; i++ { // 52 bits is maximum size of float
+			if val == math.Trunc(val) {
+				break
+			}
+			val = val * 2
+			exponent++
+		}
+		// get the mantissa
+		mantissa = int64(math.Trunc(val))
+		// mantissa can't be negative
+		if mantissa < 0 {
+			mantissa = -mantissa
+		}
+		negativeExponent = true
+		log.Debugf("Obtained mantissa is %v, value in computations is %v. Exponent is %v, it is negative (%v)", mantissa, val, exponent, negativeExponent)
+		log.Debugf("Computing 2's complement for exponent (%v)", exponent)
+		//alternative way - works for 8 bit representation numbers
+		twosComplimentExp := 256 - exponent
+		log.Debugf("2's complement for exponent %v is %v", exponent, twosComplimentExp)
+		exponent = twosComplimentExp
+		log.Debugf("It requires %v bits to store the value", howManyBitsNeeded(exponent))
+	}
+
+	p = howManyBytesNeeded(mantissa)
+	log.Debugf("Exponent/2 is %v, mantissa/2 is %v, exponent is %v, p is %v, mantissa is %v", exponent%2, mantissa%2, exponent, p, mantissa)
```

Comment:
you could say `Exponent%2` or `is odd` or `is even`
Not clear that `p` is the number of bytes